thrust/thrust/system/cuda/detail/merge.h

```diff
@@ -38,833 +38,158 @@ j * Copyright (c) 2016, NVIDIA CORPORATION.  All rights reserved.
 
 #if THRUST_DEVICE_COMPILER == THRUST_DEVICE_COMPILER_NVCC
 
-#  include <thrust/detail/cstdint.h>
-#  include <thrust/detail/mpl/math.h>
+#  include <cub/device/device_merge.cuh>
+
 #  include <thrust/detail/temporary_array.h>
 #  include <thrust/distance.h>
-#  include <thrust/extrema.h>
-#  include <thrust/merge.h>
+#  include <thrust/iterator/iterator_traits.h>
 #  include <thrust/pair.h>
-#  include <thrust/system/cuda/detail/cdp_dispatch.h>
-#  include <thrust/system/cuda/detail/core/agent_launcher.h>
-#  include <thrust/system/cuda/detail/core/util.h>
-#  include <thrust/system/cuda/detail/execution_policy.h>
-#  include <thrust/system/cuda/detail/par_to_seq.h>
+#  include <thrust/system/cuda/detail/dispatch.h>
 #  include <thrust/system/cuda/detail/util.h>
 
 THRUST_NAMESPACE_BEGIN
 namespace cuda_cub
 {
-
-namespace __merge
-{
-
-template <class KeysIt1, class KeysIt2, class Size, class BinaryPred>
-Size THRUST_DEVICE_FUNCTION
-merge_path(KeysIt1 keys1, KeysIt2 keys2, Size keys1_count, Size keys2_count, Size diag, BinaryPred binary_pred)
-{
-  using key1_type = typename iterator_traits<KeysIt1>::value_type;
-  using key2_type = typename iterator_traits<KeysIt2>::value_type;
-
-  Size keys1_begin = thrust::max<Size>(0, diag - keys2_count);
-  Size keys1_end   = thrust::min<Size>(diag, keys1_count);
-
-  while (keys1_begin < keys1_end)
-  {
-    Size mid       = (keys1_begin + keys1_end) >> 1;
-    key1_type key1 = keys1[mid];
-    key2_type key2 = keys2[diag - 1 - mid];
-    bool pred      = binary_pred(key2, key1);
-    if (pred)
-    {
-      keys1_end = mid;
-    }
-    else
-    {
-      keys1_begin = mid + 1;
-    }
+_CCCL_EXEC_CHECK_DISABLE
+template <class Derived, class KeysIt1, class KeysIt2, class ResultIt, class CompareOp = less<>>
+ResultIt _CCCL_HOST_DEVICE
+merge(execution_policy<Derived>& policy,
+      KeysIt1 keys1_begin,
+      KeysIt1 keys1_end,
+      KeysIt2 keys2_begin,
+      KeysIt2 keys2_end,
+      ResultIt result_begin,
+      CompareOp compare_op = {})
+{
+  using size_type         = typename iterator_traits<KeysIt1>::difference_type;
```

Comment:
critical: you dropped `THRUST_CDP_DISPATCH` that used to be on Thrust end for merge and merge by key. The context is the following: Thrust is synchronous, so it has to synchronize device after launching any kernels. Besides that, Thrust can be used on device side with CDP = CUDA Dynamic Parallelism, which is the ability to launch kernels from CUDA kernels. Recently, CDP dropped the ability to synchronize children kernels, so Thrust can no longer satisfy the contract of being synchronous while used on device. To keep this guarantee, Thrust switches to sequential execution when used on device. `THRUST_CDP_DISPATCH` is a helper macro telling us when we should switch to sequential backend. You can think of it as of NV_IF_TARGET.

Reply:
Thank you for the explanation, this makes sense! I readded the `THRUST_CDP_DISPATCH`.